Fixes to example/stateful_lstm.py

shadiakiki1986 opened the pull request:
- rename file to `lstm_stateful.py` so it shows up together with the other `lstm_*.py` example files
- fix input average: the existing code is not doing any averaging because `cos[i + 1:i + lahead + 1]` is just one point. I think the purpose was for it to be a 2-point average.
- aesthetic progress fix: the progress message was 0-based, i.e. first message was `Epoch 0 / 25`.

shadiakiki1986 commented:
Btw I don't think the cosine is a good example, even with the 2-point averaging.
Even if the LSTM only learns to use the current value as a prediction,
(instead of learning to average the two points),
its mean squared error would be very low.
A better example would be to replace

    cos = gen_cosine_amp()

with something like

    cos = np.random.uniform(-100, +100, (int(50e3), 1, 1))

In this case, the stateful LSTM reaches

    Epoch 25 / 25 - 50000/50000 - 4s - loss: 835.5292

whereas the stateless counterpart can reach

    Epoch 4/25 - 25000/25000 - 3s - loss: 8.4869     

I think this shows that maybe something needs to be fixed in this model with this particular example

fchollet commented:
> Btw I don't think the cosine is a good example, even with the 2-point averaging.

I agree. Feel free to replace this example script with something new entirely, that would do a better job at demonstrating stateful RNNs.

shadiakiki1986 commented:
So I replaced the cosine with random uniform data.
I also added a stateful model to show a comparison of performance between the two.
Now this example demonstrates a case that stateful can converge to a good fit whereas stateless cannot.

shadiakiki1986 commented:
I also modified the code again so that it is possible that
the user modify the `lahead` and `tsteps` in the script head
in order to make the stateless LSTM.

For example, with `tsteps=2` (i.e. 2-point average) and `lahead=1` (i.e. LSTM not seeing except 1 point), the stateful LSTM can converge whereas the stateless LSTM cannot.
On the other hand, with `tsteps=2` and `lahead=2` (i.e. LSTM sees 2 points), both stateful and stateless LSTMs can converge.

@fchollet what do you think?

fchollet commented on examples/lstm_stateful.py:
```diff
@@ -0,0 +1,206 @@
+              batch_size=batch_size,
+              return_sequences=False,
+              stateful=stateful,
+              activation='tanh'))
```
You can remove `activation` and `return_sequences` since they take the default values.

fchollet commented on examples/lstm_stateful.py:
```diff
@@ -0,0 +1,206 @@
+
+# ----------------------------
+# DO NOT EDIT UNDER THIS LINE
+# ----------------------------
```
Example scripts are meant to be edited, please remove.

fchollet commented on examples/lstm_stateful.py:
```diff
@@ -0,0 +1,206 @@
+    # tweak to match with batch_size
+    to_train -= to_train % batch_size
+
+    X_train = X[:to_train]
```
Please use lowercase variable names.

shadiakiki1986 commented:
Done. Instead of removing "DO NOT EDIT", I replaced it with "MAIN PROGRAM"

fchollet commented on examples/lstm_stateful.py:
```diff
@@ -0,0 +1,204 @@
+# -------------------
+
+# timesteps to use in the output averaging of the input
+# e.g. use 2 for 2-point average
```
Please add more info, it is not possible to understand this comment without more context.

fchollet commented on examples/lstm_stateful.py:
```diff
@@ -0,0 +1,204 @@
+
+# number of elements ahead that are used to make the LSTM prediction
+# when lahead >= tsteps, the stateless LSTM can converge
+# similar to the stateful LSTM
```
Same, please add more info. Maybe a concrete example would help.

fchollet commented on examples/lstm_stateful.py:
```diff
@@ -0,0 +1,204 @@
+'''Example script showing how to use stateful RNNs
+to model long sequences efficiently.
+Depending on the parameters, either only the stateful LSTM converges,
+or both the stateful and stateless LSTM converge.
```
Please flesh out the docstring by explaining what the example is about: what the task is, what the data look like, how the model is set up. Users should be able to go into the code already understanding what they're going to be looking at.

fchollet commented on examples/lstm_stateful.py:
```diff
@@ -0,0 +1,204 @@
+if lahead >= tsteps:
+    print("STATELESS LSTM WILL ALSO CONVERGE")
+else:
+    print("STATELESS LSTM WILL NOT  CONVERGE")
```
Typo: two spaces after NOT

fchollet commented on examples/lstm_stateful.py:
```diff
@@ -0,0 +1,204 @@
+print("*" * 33)
+
+# add a few points to account for the
+# nan-values that will be dropped in preprocessing
```
What does this mean?

shadiakiki1986 replied:
I extended the comment to explain as follows:

Since the output is a moving average of the input,
the first few points of output will be NaN
and will be dropped from the generated data
before training the LSTM.
Also, when lahead > 1, the preprocessing step later of "rolling window view"
will also cause some points to be lost.
For aesthetic reasons,
in order to maintain generated data length = input_len after pre-processing,
add a few points to account for the values that will be lost.

fchollet commented on examples/lstm_stateful.py:
```diff
@@ -0,0 +1,204 @@
+
+
+from keras.models import Sequential
+from keras.layers import Dense, LSTM
```
Put the imports at the top of the file.

shadiakiki1986 commented:
Done
